### PR TITLE
implementation of the Maliar method training loop with resampling of state spaces.

### PR DESCRIPTION
**Pull request by sbenthall**
Addresses #47 

This PR works towards a complete implementation of the Maliar method (MMW JME '21) using:
 - generic Blocks
 - sampling of the state space that converges on the ergodic distribution, for the computation of the empirical loss function
 - reasonable parameters values for the training loop

I'm sticking with the estimated lifetime reward function for this PR; swapping in alternative loss functions is out of scope.
Though I hope to make the code general enough so that's easy.

I significant part of making this work is extending the Grid class to make it easier to extend or modify (immutably -- making a copy with changes). Constructing new training sets with state and shock samples is a key part of the method. These changes are in skagent.grid and have automated tests. This class is looking more and more like a `tensordict`, but I've taken pains to make it compatible with a pure numpy implementation. It's possible that tighter integration with `tensordict` is called for and would be useful down the line.

Some of the kludgy parts of the implementation would be more smoothly done if the model analyzer #38 were in the library, since that way 'state variables' could be isolated from the DBlock's other variables through algorithmic model inspection.

### CODE REVIEW

**Comment by sbenthall** [comment]
@alanlujan91 I believe this is ready for review.

There is one currently passing test for the loop that shows the basic functionality, as well as showing that the state space does converge on an ergodic distribution.

In addition to code review, I particularly welcome ideas for automated test cases where the solution and ergodic distribution are known. Maybe there's a canonical example from consumption saving that would be useful. Note that I'm using a pretty puny estimated lifetime reward function in the cases thus far, though.

**Review by alanlujan91 — APPROVED** [review]
This looks great to me and should be merged asap. I am working on a set of "known" benchmark solutions, which would be good to start checking against
